### DISCUSSION TITLE
Can't figure out what API rule to use

### DISCUSSION THREAD

**Discussion by LickABrick** [pocketbase/pocketbase#4234] [Q&A] [answered]
Hi,

I have a multi tenant app where users can create roles, each role has permissions.
There is a collection for 'Roles' and 'Permissions'.

The roles collection basically contains a record for each collections permission type e.g.:
```
collection.List
collection.View
collection.Create
collection.Update
collection.Delete
```

- Roles are assigned to users by a relation field in the Users collection, a user can have multiple roles
- A user can be part of multiple tenants

The Roles collection looks like this:

<table>
<tr>
 <th>name (string)
 <th>tenant (single relation to tenants collection)
 <th>permissions (multi relation to permissions collection)
<tr>
 <td> Viewer
 <td> myTenant
 <td> ['collection.List', 'collection.View']
<tr>
 <td> Admin
 <td> someOtherTenant
 <td> ['collection.List', 'collection.View', 'collection.Create', 'collection.Update', 'collection.Delete' ]
</table>

The 'collection' collection looks like this:
<table>
<tr>
 <th>name
 <th>tenant (single relation to tenants collection)
 <th>someImportantData
<tr>
 <td> record 1
 <td> myTenant
 <td> really important data
<tr>
 <td> record 2
 <td> someOtherTenant
 <td> some more important data
</table>

Let's say a user is a member of both tenants mention above (`myTenant` and `someOtherTenant`)
The use has both the `Viewer` role (which belongs to `myTenant`) and the `Admin` role (which belongs to `someOtherTenant`)

The idea is the user can only delete records in the `collection` collection from `someOther` tenant and not `myTenant` as he only has `collection.View` and `collection.List` permissions in `myTenant`

How can I make sure this user can only delete records in the `collection` collection which he has the corresponding permissions for? I can't figure out what API rule to use.

(if I know what to do for Delete I also know what do for the other permission)

**Comment by ganigeorgiev** [comment, marked as answer]
There is no info about the names of the "roles" relation field in the users collection and the "permissions" collection fields.

I'm not sure that I understand your description, but if we assume that the "roles" relation field is named `roles` and the `'collection.Delete'` value is stored under a `name` field in the "permissions" collection, then the following rule should suffice:

```js
// require to be authenticated
@request.auth.id != "" &&
// the authenticated user must have at least one role with tenant matching the "collection.tenant"
@request.auth.roles.tenant ?= tenant &&
// the matched auth user role must have at least one permission with name = "collection.Delete"
@request.auth.roles.permissions.name ?= "collection.Delete"
```

If that doesn't help, please provide more clear description about the available collections and their fields (you can also export your collections configuration as json from the _Admin UI > Settings > Export collections_).

**Reply by LickABrick** [reply]
No clue why I didn't use screenshots in my initial post.. but here you can see what my collections look like:

Users collection:
![image](https://github.com/pocketbase/pocketbase/assets/17205583/df86a09e-03c3-45f0-9a08-33a0a0bb70f9)

Permissions collection:
![image](https://github.com/pocketbase/pocketbase/assets/17205583/f788e6bc-e114-439f-bc38-cfa3a8ee9d01)
(You are correct, the name field holds values like 'collection.Delete' where collection is replaced with the name of the collection)

Roles collection:
![image](https://github.com/pocketbase/pocketbase/assets/17205583/b2500383-5954-4dd5-90a0-b76763a1c52d)

With the rules you provided a User with these Tenants and Roles:
![image](https://github.com/pocketbase/pocketbase/assets/17205583/70732d27-9960-445d-ae8e-ccc113af373c)
Viewer = Role in Tenant 2
Admin = Role in Tenant 1

![image](https://github.com/pocketbase/pocketbase/assets/17205583/e11d36f1-20dc-43cf-8662-0fed664c7701)


Would still be able to delete records in the 'dataGroups' collection from Tenant 2 since the API rule would still match.

I hope this explains it a little better

**Reply by ganigeorgiev** [reply]
> Would still be able to delete records in the 'dataGroups' collection from Tenant 2 since the API rule would still match.

Did you test it? Note that in the above rule we are requiring the `roles.tenant` (and not the `users.tenants`) to match with the `dataGroups.tenant`. Or maybe I just don't understand the problem.

**Reply by ganigeorgiev** [reply]
I've tried to reproduce your data from the screenshots and the generated query seems correct to me. Here is a simplified version of it:

```sql
SELECT DISTINCT `dataGroups`.*
FROM `dataGroups`
LEFT JOIN `users` `__auth_users` ON __auth_users.id='userId'
LEFT JOIN `roles` `__auth_users_roles` ON __auth_users_roles.id [[IN __auth_users.roles]]
LEFT JOIN `permissions` `__auth_users_roles_permissions` ON __auth_users_roles_permissions.id [[IN __auth_users_roles.permissions]]
WHERE (dataGroups.id='groupId1') AND (
    (
      'userId' IS NOT ''
      AND
      __auth_users_roles.tenant = dataGroups.tenant
      AND
      __auth_users_roles_permissions.name = 'dataGroups.delete'
    )
)
LIMIT 1
```

When attempting to delete the "dataGroups" record that has "Tenant2" it fails with 404.
When attempting to delete the "dataGroups" record that has "Tenant1" it return 204.

**Reply by LickABrick** [reply]
I shouldn't doubt you, it works :)

I don't quite understand why though. Here's how I understand it using the same setup as in above screenshot with the following record:

![image](https://github.com/pocketbase/pocketbase/assets/17205583/586ecde0-6adc-442f-8b8c-58c8f97c7421)


```
// require to be authenticated
@request.auth.id != "" && --> This one is obvious.
// the authenticated user must have at least one role with tenant matching the "dataGroups.tenant"
@request.auth.roles.tenant ?= tenant && --> This one matches with the 'Viewer' role assigned to the user as the Viewer role is part of Tenant 2
// the matched auth user role must have at least one permission with name = "dataGroups.Delete"
@request.auth.roles.permissions.name ?= "dataGroups.Delete" --> This one matches the 'Admin' role since this role contains the 'dataGroups.Delete' permissions
```

I am obviously misunderstanding something but don't know what.

**Reply by ganigeorgiev** [reply]
I don't understand from where the confusion comes from.

Maybe your misunderstanding comes from the fact that you are not taking the cartesian product of the relations (the JOIN statements) in consideration?

Keep in mind that all rules execute as a **single** SQL statement, meaning that all 3 rules must be satisfied at the same time in order the action to evaluate to true.

I'm sorry, I don't know how to explain it better.

**Reply by ganigeorgiev** [reply]
Maybe a better analogy would be to imagine the rules execution as a "funnel", where with each rule we narrow down the result.

If we follow this then your third conclusion _"This one matches the 'Admin' role since..."_ would be incorrect because with the 3rd rule we will check only for the permissions within the _previously filtered/matched_ role(s).

**Reply by LickABrick** [reply]
Thanks that makes it more clear!!
